[BUG] Visualization only displaying in 1/4 of window size on macOS, python 3.8

![image](https://user-images.githubusercontent.com/1263313/132789913-147f4c6c-abdf-49ac-af1c-d899994af969.png)

Problem persists on restart.

<details>
<summary>Terminal</summary>
<pre>
(pyme) MWC02KC2D4FFRR:~ zrc4$ pymevis
/opt/miniconda3/envs/pyme/lib/python3.8/_collections_abc.py:702: MatplotlibDeprecationWarning: The global colormaps dictionary is no longer considered public API.
  return len(self._mapping)
/opt/miniconda3/envs/pyme/lib/python3.8/_collections_abc.py:720: MatplotlibDeprecationWarning: The global colormaps dictionary is no longer considered public API.
  yield from self._mapping
ERROR:PYME.recipes.modules:Could not import skimage
Traceback (most recent call last):
  File "/Users/zrc4/Code/python-microscopy/PYME/recipes/modules.py", line 26, in <module>
    from . import skfilters
  File "/Users/zrc4/Code/python-microscopy/PYME/recipes/skfilters.py", line 90, in <module>
    a = args[-j]
IndexError: list index out of range
Trying to load 3rd party recipe module ch_shrinkwrap.recipe_modules.surface_fitting
Loaded 3rd party recipe module ch_shrinkwrap.recipe_modules.surface_fitting
quaternion module not found, disabling custom clip plane orientations
quaternion module not found, disabling custom clip plane orientations
New Canvas
DEBUG:PYME.LMVis.visCore:Creating VisGUI menu bar
DEBUG:PYME.LMVis.Extras:Initializing QPobjectSegment plugin
DEBUG:PYME.LMVis.Extras:Initializing animation plugin
DEBUG:PYME.LMVis.Extras:Initializing clusterAnalysis plugin
DEBUG:PYME.LMVis.Extras:Initializing dockedPanel plugin
DEBUG:PYME.LMVis.Extras:Initializing extra_layers plugin
DEBUG:PYME.LMVis.Extras:Initializing fiducial plugin
DEBUG:PYME.LMVis.Extras:Initializing flagZMotionArtifacts plugin
DEBUG:root:Adding menu items for event filters
DEBUG:PYME.LMVis.Extras:Initializing fusionUtils plugin
DEBUG:PYME.LMVis.Extras:Initializing multiColourAnalysis plugin
DEBUG:PYME.LMVis.Extras:Initializing multiviewMapping plugin
DEBUG:root:Adding menu items for multi-view manipulation
DEBUG:PYME.LMVis.Extras:Initializing objectMeasurements plugin
DEBUG:PYME.LMVis.Extras:Initializing particleTracking plugin
DEBUG:PYME.LMVis.Extras:Initializing perFrameVariable plugin
DEBUG:PYME.LMVis.Extras:Initializing photophysics plugin
DEBUG:PYME.LMVis.Extras:Initializing pointSetGeneration plugin
DEBUG:PYME.LMVis.Extras:Initializing pointwiseColoc plugin
DEBUG:PYME.LMVis.Extras:Initializing saved_views plugin
Saved Views are deprecated -  use the ones with animation instead and remove me!
DEBUG:PYME.LMVis.Extras:Initializing scaleBox plugin
DEBUG:PYME.LMVis.Extras:Initializing sequentialSpecies plugin
DEBUG:PYME.LMVis.Extras:Initializing setBGImage plugin
DEBUG:PYME.LMVis.Extras:Initializing shiftmapGenerator plugin
DEBUG:PYME.LMVis.Extras:Initializing showShortcuts plugin
DEBUG:PYME.LMVis.Extras:Initializing snapshot plugin
DEBUG:PYME.LMVis.Extras:Initializing spherical_harmonics plugin
DEBUG:root:Adding menu items for spherical harmonic (shell) fitting
DEBUG:PYME.LMVis.Extras:Initializing subsets plugin
DEBUG:PYME.LMVis.Extras:Initializing surface_fitting plugin
DEBUG:PYME.LMVis.Extras:Initializing temperaturePlot plugin
DEBUG:PYME.LMVis.Extras:Initializing vibration plugin
Creating fold panel
GenPanels
Creating datasource panel
RecipeView._layout
Input datasources:  []
Image list size: 1
Creating layers GUI
height:  0
NB best size: wx.Size(200, 0)
RecipeView._layout
Input datasources:  []
recipe.execute()
RecipeView._layout
Input datasources:  []
0
OpenGL - Version: b'2.1 INTEL-14.7.18'
Shader - Version: b'1.20'
GL_MAX_SAMPLES: 8, GL_SAMPLES: 4
DEBUG:PYME.LMVis.shader_programs.ShaderProgramFactory:New shader program created: <class 'PYME.LMVis.shader_programs.DefaultShaderProgram.DefaultShaderProgram'>
Ev Idle
('points', 't')
refreshed
Creating Pipeline
RecipeView._layout
10:31:55 PM: Debug: Adding duplicate image handler for 'Windows bitmap file'
10:31:55 PM: Debug: Adding duplicate animation handler for '1' type
10:31:55 PM: Debug: Adding duplicate animation handler for '2' type
10:31:55 PM: Debug: Adding duplicate image handler for 'Windows bitmap file'
10:31:55 PM: Debug: Adding duplicate animation handler for '1' type
10:31:55 PM: Debug: Adding duplicate animation handler for '2' type
Input datasources:  ['_Localizations', 'FitResults']
RecipeView._layout
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/common/wincmn.cpp(470) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
Input datasources:  ['_Localizations', 'FitResults']
RecipeView._layout
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/common/wincmn.cpp(470) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/common/wincmn.cpp(470) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
Input datasources:  ['_Localizations', 'FitResults']
recipe.execute()
Executing <PYME.recipes.localisations.Pipelineify object at 0x7fcd55808360>
DEBUG:PYME.recipes.localisations:no events found
Executing <PYME.recipes.localisations.ProcessColour object at 0x7fcd592f3540>
Executing <PYME.recipes.tablefilters.FilterTable object at 0x7fcd595c8220>
RecipeView._layout
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/common/wincmn.cpp(470) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/common/wincmn.cpp(470) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/common/wincmn.cpp(470) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
Input datasources:  ['_Localizations', 'FitResults']
Pipeline Created
RecipeView._layout
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/common/wincmn.cpp(470) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/common/wincmn.cpp(470) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
wx._core.wxAssertionError: C++ assertion "GetEventHandler() == this" failed at /Users/robind/projects/bb2/dist-osx-py38/build/ext/wxWidgets/src/common/wincmn.cpp(470) in ~wxWindowBase(): any pushed event handlers must have been removed

The above exception was the direct cause of the following exception:

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
Input datasources:  ['_Localizations', 'FitResults']
DEBUG:PYME.LMVis.VisGUI:Creating tabs
0
DEBUG:PYME.LMVis.VisGUI:Finished creating tabs
Gui stuff done
DEBUG:PYME.LMVis.shader_programs.ShaderProgramFactory:New shader program created: <class 'PYME.LMVis.shader_programs.DefaultShaderProgram.OpaquePointShaderProgram'>
lw update
Creating layers GUI
hl update
height:  308
NB best size: wx.Size(235, 308)
DEBUG:PYME.LMVis.visCore:Added layer, datasouce=output
lw update
hl update
hl update
</pre>
</details>

<details>
<summary>Package list</summary>
<pre>
# packages in environment at /opt/miniconda3/envs/pyme:
#
# Name                    Version                   Build  Channel
alabaster                 0.7.12                   pypi_0    pypi
aniso8601                 9.0.1                    pypi_0    pypi
babel                     2.9.1                    pypi_0    pypi
ca-certificates           2021.7.5             hecd8cb5_1  
certifi                   2021.5.30        py38hecd8cb5_0  
ch-shrinkwrap             0.0                       dev_0    <develop>
charset-normalizer        2.0.4                    pypi_0    pypi
cheroot                   8.5.2                    pypi_0    pypi
click                     8.0.1                    pypi_0    pypi
cycler                    0.10.0                   pypi_0    pypi
cython                    0.29.24                  pypi_0    pypi
docutils                  0.17.1                   pypi_0    pypi
enum-compat               0.0.3                    pypi_0    pypi
flask                     1.1.2                    pypi_0    pypi
flask-restful             0.3.9                    pypi_0    pypi
flask-sqlalchemy          2.5.1                    pypi_0    pypi
greenlet                  1.1.1                    pypi_0    pypi
idna                      3.2                      pypi_0    pypi
imageio                   2.9.0                    pypi_0    pypi
imagesize                 1.2.0                    pypi_0    pypi
importlib-metadata        4.6.4                    pypi_0    pypi
importlib-resources       5.2.2                    pypi_0    pypi
itsdangerous              2.0.1                    pypi_0    pypi
jaraco-classes            3.2.1                    pypi_0    pypi
jaraco-collections        3.4.0                    pypi_0    pypi
jaraco-functools          3.3.0                    pypi_0    pypi
jaraco-text               3.5.1                    pypi_0    pypi
jinja2                    3.0.1                    pypi_0    pypi
joblib                    1.0.1                    pypi_0    pypi
kiwisolver                1.3.1                    pypi_0    pypi
libcxx                    10.0.0                        1  
libffi                    3.3                  hb1e8313_2  
markupsafe                2.0.1                    pypi_0    pypi
marshmallow               3.12.1                   pypi_0    pypi
matplotlib                3.4.3                    pypi_0    pypi
more-itertools            8.8.0                    pypi_0    pypi
mpld3                     0.5.5                    pypi_0    pypi
mysql-connector-python    8.0.25                   pypi_0    pypi
ncurses                   6.2                  h0a44026_1  
netifaces                 0.11.0                   pypi_0    pypi
networkx                  2.6.2                    pypi_0    pypi
numexpr                   2.7.3                    pypi_0    pypi
numpy                     1.21.2                   pypi_0    pypi
openssl                   1.1.1k               h9ed2024_0  
packaging                 21.0                     pypi_0    pypi
pandas                    1.3.2                    pypi_0    pypi
pillow                    8.3.1                    pypi_0    pypi
pip                       21.0.1           py38hecd8cb5_0  
portend                   2.7.1                    pypi_0    pypi
protobuf                  3.17.3                   pypi_0    pypi
pyface                    7.1.0                    pypi_0    pypi
pyfftw                    0.12.0                   pypi_0    pypi
pygments                  2.10.0                   pypi_0    pypi
pyopengl                  3.1.5                    pypi_0    pypi
pyparsing                 2.4.7                    pypi_0    pypi
python                    3.8.11               h88f2d9e_1  
python-dateutil           2.8.2                    pypi_0    pypi
python-microscopy         21.7.7                    dev_0    <develop>
python.app                2                       py38_10  
pytz                      2021.1                   pypi_0    pypi
pywavelets                1.1.1                    pypi_0    pypi
pyyaml                    5.4.1                    pypi_0    pypi
readline                  8.1                  h9ed2024_0  
requests                  2.26.0                   pypi_0    pypi
scikit-image              0.18.2                   pypi_0    pypi
scikit-learn              0.24.2                   pypi_0    pypi
scipy                     1.7.1                    pypi_0    pypi
setuptools                52.0.0           py38hecd8cb5_0  
six                       1.16.0                   pypi_0    pypi
snowballstemmer           2.1.0                    pypi_0    pypi
sphinx                    4.1.2                    pypi_0    pypi
sphinxcontrib-applehelp   1.0.2                    pypi_0    pypi
sphinxcontrib-devhelp     1.0.2                    pypi_0    pypi
sphinxcontrib-htmlhelp    2.0.0                    pypi_0    pypi
sphinxcontrib-jsmath      1.0.1                    pypi_0    pypi
sphinxcontrib-qthelp      1.0.3                    pypi_0    pypi
sphinxcontrib-serializinghtml 1.1.5                    pypi_0    pypi
sqlalchemy                1.4.15                   pypi_0    pypi
sqlite                    3.36.0               hce871da_0  
tables                    3.6.1                    pypi_0    pypi
tempora                   4.1.1                    pypi_0    pypi
threadpoolctl             2.2.0                    pypi_0    pypi
tifffile                  2021.8.8                 pypi_0    pypi
tk                        8.6.10               hb0a8c7a_0  
toposort                  1.6                      pypi_0    pypi
traits                    6.1.0                    pypi_0    pypi
traitsui                  7.1.0                    pypi_0    pypi
ujson                     4.1.0                    pypi_0    pypi
urllib3                   1.26.6                   pypi_0    pypi
werkzeug                  2.0.1                    pypi_0    pypi
wheel                     0.37.0             pyhd3eb1b0_0  
wxpython                  4.1.1                    pypi_0    pypi
xz                        5.2.5                h1de35cc_0  
zc-lockfile               2.0                      pypi_0    pypi
zeroconf                  0.17.7                   pypi_0    pypi
zipp                      3.5.0                    pypi_0    pypi
zlib                      1.2.11               h1de35cc_3 
</pre>
</details>

--- a/PYME/IO/DataSources/CropDataSource.py
+++ b/PYME/IO/DataSources/CropDataSource.py
@@ -155,6 +155,7 @@ class DataSource(XYZTCDataSource):
 
 def crop_image(image, xrange=None, yrange=None, zrange=None, trange=None):
     from PYME.IO.image import ImageStack
+    from PYME.IO.DataSources import ArrayDataSource
 
     vx, vy, vz = image.voxelsize
     ox, oy, oz = image.origin
@@ -167,7 +168,15 @@ def crop_image(image, xrange=None, yrange=None, zrange=None, trange=None):
         else:
             return r[0]
 
-    cropped = DataSource(image.data_xyztc, xrange=xrange, yrange=yrange, zrange=zrange, trange=trange)
+    if isinstance(image.data_xyztc, ArrayDataSource.XYZTCArrayDataSource):
+        # special case for array data source so that we can use views and get optimal chunk access (dask/zarr)
+        #  - use array slicing rather slice slicing
+        # TODO - change CropDataSource to be chunk aware
+        s = image.data_xyztc.shape
+        cp_a = image.data_xyztc[DataSource._sliceify(xrange, s[0]), DataSource._sliceify(yrange, s[1]), DataSource._sliceify(zrange, s[2]), DataSource._sliceify(trange, s[3]), :]
+        cropped = ArrayDataSource.XYZTCArrayDataSource(cp_a)
+    else:
+        cropped = DataSource(image.data_xyztc, xrange=xrange, yrange=yrange, zrange=zrange, trange=trange)
     
     im = ImageStack(cropped, titleStub='Cropped Image')
     im.mdh.copyEntriesFrom(image.mdh)

--- a/PYME/LMVis/gl_render3D_shaders.py
+++ b/PYME/LMVis/gl_render3D_shaders.py
@@ -290,8 +290,19 @@ class LMGLShaderCanvas(GLCanvas):
     def interlace_stencil(self):
         window_width = self.view_port_size[0]
         window_height = self.view_port_size[1]
+
+        #high dpi screens
+        # fix scaling error
+        # TODO - should this check be on wx, or OpenGL (or potentially python)
+        # TODO - do we need to do anything differnt on win?
+        vmajor,vminor = wx.VERSION[:2]
+        if vmajor > 4 and vminor >=1: # 4.1 or later 
+            sc = self.GetContentScaleFactor()
+        else:
+            sc = 1
+
         # setting screen-corresponding geometry
-        GL.glViewport(0, 0, window_width, window_height)
+        GL.glViewport(0, 0, int(sc*window_width), int(sc*window_height))
         GL.glMatrixMode(GL.GL_MODELVIEW)
         GL.glLoadIdentity()
         GL.glMatrixMode(GL.GL_PROJECTION)
@@ -452,8 +463,17 @@ class LMGLShaderCanvas(GLCanvas):
         #print('resize_gl')
         self._oit_w = w
         self._oit_h = h
+
+        # fix scaling error
+        # TODO - should this check be on wx, or OpenGL (or potentially python)
+        # TODO - do we need to do anything differnt on win?
+        vmajor,vminor = wx.VERSION[:2]
+        if vmajor > 4 and vminor >=1: # 4.1 or later 
+            sc = self.GetContentScaleFactor()
+        else:
+            sc = 1
     
-        GL.glViewport(0, 0, w, h)
+        GL.glViewport(0, 0, int(sc*w), int(sc*h))
         #self._accumdata = np.zeros([4,w,h], 'f')
     
         #print('bind fb')
@@ -791,6 +811,7 @@ class LMGLShaderCanvas(GLCanvas):
 
     def _ScreenCoordinatesToNm(self, x, y):
         # FIXME!!!
+        sc = self.GetContentScaleFactor()
         x_ = self.pixelsize * (x - 0.5 * float(self.view_port_size[0])) + self.view.translation[0]
         y_ = self.pixelsize * (y - 0.5 * float(self.view_port_size[1])) + self.view.translation[1]
         # print x_, y_

